### ISSUE TITLE
nan Loss with MTP Loss

### ISSUE THREAD

**Issue by semirelezovikj** [nutonomy/nuscenes-devkit#520]
Training using MTP yields nan MTPLoss - do you have a training example where you circumvent this problem? Any suggestions are welcome. 

Details below: 

```
import torch
import torch.optim as optim
import torch.nn as nn
from torch.utils.data import DataLoader, Dataset
from typing import List
from nuscenes.prediction.input_representation.static_layers import StaticLayerRasterizer
from nuscenes.prediction.input_representation.agents import AgentBoxesWithFadedHistory
from nuscenes.prediction.input_representation.interface import InputRepresentation
from nuscenes.prediction.input_representation.combinators import Rasterizer
from nuscenes.prediction.models.backbone import ResNetBackbone
from nuscenes.prediction.models.mtp import MTP
from nuscenes.prediction.models.covernet import CoverNet
from nuscenes.prediction import PredictHelper
import numpy as np

helper = PredictHelper(nuscenes)

static_layer_rasterizer = StaticLayerRasterizer(helper)
agent_rasterizer = AgentBoxesWithFadedHistory(helper, seconds_of_history=1)
mtp_input_representation = InputRepresentation(static_layer_rasterizer, agent_rasterizer, Rasterizer())

# torch.set_default_tensor_type('torch.FloatTensor') # default tensor is cuda


class MTPDataset(Dataset):
    def __init__(self, tokens: List[str], helper: PredictHelper):
        self.tokens = tokens
        self.static_layer_representation = StaticLayerRasterizer(helper)
        
    def __len__(self):
        return len(self.tokens)

    def __getitem__(self, index: int):

        token = self.tokens[index]
        instance_token, sample_token = token.split("_")

        image = self.static_layer_representation.make_representation(instance_token, sample_token)
        image = torch.Tensor(image).permute(2, 0, 1)
        
        agent_state_vector = torch.Tensor([helper.get_velocity_for_agent(instance_token, sample_token),
                                        helper.get_acceleration_for_agent(instance_token, sample_token),
                                        helper.get_heading_change_rate_for_agent(instance_token, sample_token)])
        
        ground_truth = helper.get_future_for_agent(instance_token, sample_token, 6, in_agent_frame=False)
        ground_truth = np.expand_dims(ground_truth, 0)

        return image, agent_state_vector, ground_truth
```

```
import torchvision

from nuscenes.prediction.models.mtp import MTP, MTPLoss
NUM_MODES = 2
backbone = ResNetBackbone('resnet50')
model = MTP(backbone, NUM_MODES)

device = torch.device('cuda:0' if torch.cuda.is_available() else 'cpu')
print('running on ', device)
model = model.to(device)

if torch.cuda.device_count() > 1:
  print("Let's use", torch.cuda.device_count(), "GPUs!")
  model = nn.DataParallel(model)

tokens = full_train[:100]

dataset = MTPDataset(tokens, helper)
dataloader = DataLoader(dataset, batch_size=16*torch.cuda.device_count(), num_workers=4 * torch.cuda.device_count())

n_iter = 0

minimum_loss = 0

loss_function = MTPLoss(NUM_MODES, 1, 5)

current_loss = 10000

optimizer = optim.SGD(model.parameters(), lr=0.1)


for img, agent_state_vector, ground_truth in dataloader:
    print('Iteration', n_iter, 'img.shape', img.shape, 'agent_state_vector.shape', agent_state_vector.shape)
    imshow(torchvision.utils.make_grid(img)) 
    img = img.to(device)
    agent_state_vector = agent_state_vector.to(device)    
    ground_truth = ground_truth.to(device)
    ground_truth = ground_truth.to(torch.float32)

    optimizer.zero_grad()
    
    
    
    prediction = model(img, agent_state_vector)    

    
    print('prediction.shape', prediction.shape, 'ground_truth.shape', ground_truth.shape)
    loss = loss_function(prediction, ground_truth)
    loss.backward()
    optimizer.step()

    current_loss = loss.cpu().detach().numpy()
    
       
    n_iter += 1
    
    print(f"Current loss is {current_loss:.4f}")
    if np.allclose(current_loss, minimum_loss, atol=1e-4):
        print(f"Achieved near-zero loss after {n_iter} iterations.")
        break

        n_iter += 1
        
    
print('completed')
```

yields

```
Current loss is nan
```

**Comment by semirelezovikj** [comment]
In case anybody is wondering - here is how many `agent_state_vector` with NaN values are in the `val` and 'train' set respectively:

```
643 / 9041 for val
2297 / 32186 for train
```

```
how_many_val_nans = 0
total_val = 0
how_many_train_nans = 0
total_train = 0

for token in val:
    total_val += 1
    instance_token, sample_token = token.split("_")

    agent_state_vector = torch.Tensor([helper.get_velocity_for_agent(instance_token, sample_token),
                                        helper.get_acceleration_for_agent(instance_token, sample_token),
                                        helper.get_heading_change_rate_for_agent(instance_token, sample_token)])
    
    if agent_state_vector.isnan().sum() > 0:
        how_many_val_nans += 1

print(how_many_val_nans, '/', total_val, 'for val')

for token in full_train:
    total_train += 1
    instance_token, sample_token = token.split("_")

    agent_state_vector = torch.Tensor([helper.get_velocity_for_agent(instance_token, sample_token),
                                        helper.get_acceleration_for_agent(instance_token, sample_token),
                                        helper.get_heading_change_rate_for_agent(instance_token, sample_token)])
    
    if agent_state_vector.isnan().sum() > 0:
        how_many_train_nans += 1


print(how_many_train_nans, '/', total_train, 'for train')
```

**Comment by emwolff** [comment]
@semirelezovikj Thanks! @holger-nutonomy Let's chat after break on fixing this in a future release.
